File: system-test/java/src/main/java/com/mycompany/myshop/testkit/dsl/port/given/steps/GivenCoupon.java
package com.mycompany.myshop.testkit.dsl.port.given.steps;

import com.mycompany.myshop.testkit.dsl.port.given.steps.base.GivenStep;

public interface GivenCoupon extends GivenStep {
    GivenCoupon withCouponCode(String couponCode);

    GivenCoupon withDiscountRate(String discountRate);

    GivenCoupon withDiscountRate(double discountRate);

    GivenCoupon withValidFrom(String validFrom);

    GivenCoupon withValidTo(String validTo);

    GivenCoupon withUsageLimit(String usageLimit);

    GivenCoupon withUsageLimit(int usageLimit);
}

File: system-test/typescript/src/testkit/dsl/core/scenario/given/given-coupon.ts
import { CouponConfig } from '../scenario-context.js';
import { ThenContractStage } from '../then/then-contract.js';
import { WhenStage } from '../when/when-stage.js';
import type { GivenStage } from './given-stage.js';
import type { GivenCoupon as IGivenCoupon } from '../../../port/given/steps/given-coupon.js';

export class GivenCoupon implements IGivenCoupon {
  constructor(
    private readonly stage: GivenStage,
    private readonly config: CouponConfig,
  ) {}

  withCode(code: string): this {
    this.config.code = code;
    return this;
  }

  withCouponCode(code: string): this {
    return this.withCode(code);
  }

  withDiscountRate(discountRate: number): this {
    this.config.discountRate = discountRate;
    return this;
  }

  withValidFrom(validFrom: string): this {
    this.config.validFrom = validFrom;
    return this;
  }

  withValidTo(validTo: string): this {
    this.config.validTo = validTo;
    return this;
  }

  withUsageLimit(usageLimit: number | string): this {
    this.config.usageLimit = usageLimit;
    return this;
  }

  and(): GivenStage {
    return this.stage;
  }

  when(): WhenStage {
    return this.stage.when();
  }

  then(): ThenContractStage {
    return this.stage.then();
  }
}
